enhancements/network/baremetal-ipi-network-configuration.md

```diff
@@ -0,0 +1,279 @@
+
+RHCOS already provides a mechanism to specify NetworkManager keyfiles during deployment of a new node. We need to expose that functionality during the IPI install process. There are a couple of options on how to do that:
+
+* A new section in install-config.
+* A secret that contains base64-encoded content for the keyfiles.
+
+These are not mutually exclusive. If we implement the install-config option, we will still need to persist the configuration in a secret so it can be used for day-2. In fact, the initial implementation should likely be based on secrets, and then install-config integration can be added on top of that to improve the user experience. This way the install-config work won't block implementation of the feature.
+
+The data provided by the user will need to have the following structure:
+```yaml
+[hostname]:
+  [interface name]:
+    [contents of NetworkManager key file for interface]
+  [second interface name]:
+    [contents of NetworkManager key file for second interface]
+  etc...
+```
+
+For example:
+```yaml
+openshift-master-0:
+  eth0:
+    [connection]
+    id=eth0
+    uuid=18e0cec7-041c-4fb3-957c-a60c80dd9b85
+    type=ethernet
+    etc...
+openshift-master-1:
+  eth0:
+    [connection]
+    id=eth0
+    etc...
+```
+
+In install-config this would look like:
+```yaml
+platform:
+  baremetal:
+    hosts:
+      - name: openshift-master-0
+        networkConfig:
+          eth0:
+            [connection]
+            id=eth0
+            uuid=18e0cec7-041c-4fb3-957c-a60c80dd9b85
+            type=ethernet
+            etc...
+```
+
+Because the initial implementation will be baremetal-specific, we can put the
+network configuration data into the baremetal host field, which will allow easy
+mapping to the machine in question.
+
+#### API Open Questions
+
+* What is the best way to encode the data above into secret(s)?
+
+### Processing user configuration
+
+#### Deployment of the controlplane hosts via terraform
+
+We will map the keyfiles to their appropriate Machine using the host field
+of the baremetal install-config. The keyfiles will then be added to custom
+images for each host built by Terraform and Ironic.
+
+TODO - Does anyone have more detail to add here?
+
+#### Deployment of compute/worker hosts
+
+BareMetalHost resources for workers will be created with the Secret containing the network data referenced in the `preprovisioningNetworkData` field defined in the Metal³ [image builder integration design](https://github.com/metal3-io/metal3-docs/blob/master/design/baremetal-operator/image-builder-integration.md#custom-agent-image-controller).
+This will cause the baremetal-operator to create a PreprovisioningImage CRD and wait for it to become available before booting the IPA image.
+
+An OpenShift-specific PreprovisioningImage controller will use the provided network data to build a CoreOS IPA image with the correct network data in place, and the baremetal-operator will then use this to boot the Host.
```

Comment:
nit: s/network data/ignition configuration/ (for clarity)

Reply:
Ack, done.